db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/VerifyAggregateExpressionsVisitor.java

```diff
@@ -36,137 +36,142 @@
-	 * is equivalent to any of the group by expressions.
-	 *
-	 * @param node 	the node to process
-	 *
-	 * @return true/false
-	 * @throws StandardException 
-	 */
-	public boolean skipChildren(Visitable node) throws StandardException 
-	{
-		// skip aggregate node but not window function node
-		return ((node instanceof AggregateNode && !(node instanceof WindowFunctionNode)) ||
-				(node instanceof SubqueryNode) ||
-				(node instanceof ValueNode &&
-						groupByList != null 
-						&& groupByList.findGroupingColumn((ValueNode)node) != null));
-	}
-	
-	public boolean stopTraversal()
-	{
-		return false;
-	}
-
-	public boolean visitChildrenFirst(Visitable node)
-	{
-		return false;
-	}
+    public Visitable visit(Visitable node, QueryTreeNode parent) throws StandardException {
+        if (node instanceof ColumnReference) {
+            ColumnReference cr = (ColumnReference) node;
+
+            if (groupByList == null) {
+                throw StandardException.newException(SQLState.LANG_INVALID_COL_REF_NON_GROUPED_SELECT_LIST, cr.getSQLColumnName());
+            }
+
+            if (groupByList.findGroupingColumn(cr) == null) {
+                throw StandardException.newException(SQLState.LANG_INVALID_COL_REF_GROUPED_SELECT_LIST, cr.getSQLColumnName());
+            }
+        }
+
+        /*
+         ** Subqueries are only valid if they do not have
+         ** correlations and are expression subqueries.  RESOLVE:
+         ** this permits VARIANT expressions in the subquery --
+         ** should this be allowed?  may be confusing to
+         ** users to complain about:
+         **
+         ** select max(x), (select sum(y).toString() from y) from x
+         */
+        else if (node instanceof SubqueryNode) {
+            SubqueryNode subquery = (SubqueryNode) node;
+
+            if ((subquery.getSubqueryType() != SubqueryNode.EXPRESSION_SUBQUERY) ||
+                    subquery.hasCorrelatedCRs()) {
+                if (groupByList != null) {
+                    /*
+                     ** Check if the SSQ's correlated columns are referencing the grouping columns, if so, allow it.
+                     */
+                    List<ValueNode> correlationCRs = subquery.getCorrelationCRs();
+                    for (ValueNode correlationCR : correlationCRs) {
+                        if (correlationCR instanceof ColumnReference) {
+                            boolean included = false;
+                            for (OrderedColumn column : groupByList.getNodes()) {
+                                if (column.getValueNode().equals(correlationCR)) {
+                                    included = true;
+                                    break;
+                                }
+                            }
+                            if (!included) {
+                                throw StandardException.newException(SQLState.LANG_INVALID_GROUPED_SELECT_LIST);
+                            }
```

Comment:
Can you use `groupByList.findGroupingColumn(cr)` instead like we do above for ColumnReference?

Reply:
done.